std/atomic.d

```diff
@@ -0,0 +1,314 @@
+	{
+		return fadd(1);
+	}
+
+	/** Implicit conversion to FSUB with SC access
+	  Params:
+		op = --
+
+	  Returns: Pre-decremented value
+	*/
+	T opUnary(string op)() shared if (op == `--`)
+	{
+		return fsub(1);
+	}
+
+	/** Dereference atomic pointer with SC access
+	  Params:
+		op = *
+
+	  Returns: Reference to the pointed location
+	*/
+	auto ref opUnary(string op)() shared if (op == `*`)
+	{
+		return *(load);
+	}
+}
+
+static import core.atomic;
+/**
+ * Specifies the memory ordering semantics of an atomic operation.
+ *
+ * See_Also:
+ *     $(HTTP en.cppreference.com/w/cpp/atomic/memory_order)
+ */
+enum MemoryOrder
+{
+	/**
+	 * Corresponds to $(LINK2 https://llvm.org/docs/Atomics.html#monotonic, LLVM AtomicOrdering.Monotonic)
+	 * and C++11/C11 `memory_order_relaxed`.
+	 */
+	rlx = cast(int) core.atomic.MemoryOrder.raw,
+	/**
+	 * Corresponds to $(LINK2 https://llvm.org/docs/Atomics.html#acquire, LLVM AtomicOrdering.Acquire)
+	 * and C++11/C11 `memory_order_acquire`.
+	 */
+	acq = cast(int) core.atomic.MemoryOrder.acq,
+	/**
+	 * Corresponds to $(LINK2 https://llvm.org/docs/Atomics.html#release, LLVM AtomicOrdering.Release)
+	 * and C++11/C11 `memory_order_release`.
+	 */
+	rel = cast(int) core.atomic.MemoryOrder.rel,
+	/**
+	 * Corresponds to $(LINK2 https://llvm.org/docs/Atomics.html#acquirerelease, LLVM AtomicOrdering.AcquireRelease)
+	 * and C++11/C11 `memory_order_acq_rel`.
+	 */
+	acq_rel = cast(int) core.atomic.MemoryOrder.acq_rel,
+	/**
+	 * Corresponds to $(LINK2 https://llvm.org/docs/Atomics.html#sequentiallyconsistent, LLVM AtomicOrdering.SequentiallyConsistent)
+	 * and C++11/C11 `memory_order_seq_cst`.
+	 */
+	seq = cast(int) core.atomic.MemoryOrder.seq,
+}
+
+private auto toCore(MemoryOrder mo)
+{
+	static import core.atomic;
+	return cast(core.atomic.MemoryOrder) mo;
+}
+
+@safe unittest
+{
+	shared Atomic!int a;
+	assert(a == 0);
+	assert(a.load == 0);
+	assert(a.fadd!(MemoryOrder.rlx)(5) == 0);
+	assert(a.load!(MemoryOrder.acq) == 5);
+	assert(!a.casWeak(4, 5));
+	assert(!a.cas(4, 5));
+	assert(a.cas!(MemoryOrder.rel, MemoryOrder.acq)(5, 4));
+	assert(a.fsub!(MemoryOrder.acq_rel)(2) == 4);
+	assert(a.exchange!(MemoryOrder.acq_rel)(3) == 2);
+	assert(a.load!(MemoryOrder.rlx) == 3);
+	a.store!(MemoryOrder.rel)(7);
+	assert(a.load == 7);
+	a = 32;
+	assert(a == 32);
+	a += 5;
+	assert(a == 37);
+	assert(a++ == 37);
+	assert(a == 38);
+}
+
+// static array of shared atomics
+@safe unittest
+{
+	static shared(Atomic!int)[5] arr;
+	arr[4] = 4;
+	assert(arr[4].load == 4);
+}
+
+@system unittest
+{
+	import core.thread : Thread;
+
+	shared(Atomic!int)[2] arr;
+
+	void reltest() @safe
+	{
+		arr[0].store!(MemoryOrder.rel)(1);
+		arr[1].store!(MemoryOrder.rel)(1);
+	}
+
+	void acqtest() @safe
+	{
+		while (arr[1].load!(MemoryOrder.acq) != 1) { }
+		assert(arr[0].load!(MemoryOrder.acq) == 1);
+	}
+
+	auto t1 = new Thread(&acqtest);
+	auto t2 = new Thread(&reltest);
+	t2.start;
+	t1.start;
+	t2.join;
+	t1.join;
```

Comment:
You can't possibly test anything like this with one sample.

Reply:
The alignment needs to be for the variable instance, we don't want to align all atomics.